perception/traffic_light_multi_camera_fusion/src/node.cpp

```diff
@@ -0,0 +1,291 @@
+        ExactSyncPolicy(10), *(cam_info_subs_.back()), *(roi_subs_.back()),
+        *(signal_subs_.back())));
+      exact_sync_subs_.back()->registerCallback(
+        std::bind(&MultiCameraFusion::trafficSignalRoiCallback, this, _1, _2, _3));
+    } else {
+      appro_sync_subs_.emplace_back(new ApproSync(
+        ApproSyncPolicy(10), *(cam_info_subs_.back()), *(roi_subs_.back()),
+        *(signal_subs_.back())));
+      appro_sync_subs_.back()->registerCallback(
+        std::bind(&MultiCameraFusion::trafficSignalRoiCallback, this, _1, _2, _3));
+    }
+  }
+
+  if (perform_group_fusion_) {
+    map_sub_ = create_subscription<autoware_auto_mapping_msgs::msg::HADMapBin>(
+      "~/input/vector_map", rclcpp::QoS{1}.transient_local(),
+      std::bind(&MultiCameraFusion::mapCallback, this, _1));
+  }
+  signal_pub_ = create_publisher<SignalArrayType>("~/output/traffic_signals", rclcpp::QoS{1});
+}
+
+void MultiCameraFusion::trafficSignalRoiCallback(
+  const CamInfoType::ConstSharedPtr cam_info_msg, const RoiArrayType::ConstSharedPtr roi_msg,
+  const SignalArrayType::ConstSharedPtr signal_msg)
+{
+  rclcpp::Time stamp(roi_msg->header.stamp);
+  /*
+  Insert the received record array to the table.
+  Attention should be paied that this record array might not have the newest timestamp
+  */
+  record_arr_set_.insert(FusionRecordArr{cam_info_msg->header, *roi_msg, *signal_msg});
+
+  std::map<IdType, FusionRecord> fusionedRecordMap;
+  multiCameraFusion(*cam_info_msg, fusionedRecordMap);
+  if (perform_group_fusion_) {
+    groupFusion(*cam_info_msg, fusionedRecordMap);
+  }
+
+  SignalArrayType out_msg;
+  out_msg.header = signal_msg->header;
+  out_msg.signals.reserve(fusionedRecordMap.size());
+  for (const auto & p : fusionedRecordMap) {
+    out_msg.signals.emplace_back(p.second.signal);
+  }
+  signal_pub_->publish(out_msg);
+}
+
+void MultiCameraFusion::mapCallback(
+  const autoware_auto_mapping_msgs::msg::HADMapBin::ConstSharedPtr input_msg)
+{
+  lanelet::LaneletMapPtr lanelet_map_ptr = std::make_shared<lanelet::LaneletMap>();
+
+  lanelet::utils::conversion::fromBinMsg(*input_msg, lanelet_map_ptr);
+  lanelet::ConstLanelets all_lanelets = lanelet::utils::query::laneletLayer(lanelet_map_ptr);
+  std::vector<lanelet::AutowareTrafficLightConstPtr> all_lanelet_traffic_lights =
+    lanelet::utils::query::autowareTrafficLights(all_lanelets);
+  for (auto tl_itr = all_lanelet_traffic_lights.begin(); tl_itr != all_lanelet_traffic_lights.end();
+       ++tl_itr) {
+    lanelet::AutowareTrafficLightConstPtr tl = *tl_itr;
+
+    auto lights = tl->trafficLights();
+    for (const auto & light : lights) {
+      trafficLightId2RegulatoryEleId_[light.id()] = tl->id();
+    }
+  }
+}
+
+void MultiCameraFusion::multiCameraFusion(
+  const CamInfoType cam_info, std::map<IdType, FusionRecord> & fusionedRecordMap)
+{
+  fusionedRecordMap.clear();
+  /*
+  this should not happen. Just in case
+  */
+  if (record_arr_set_.empty()) {
```

Comment:
nitpicks:
If some condition should not happen, maybe a console message would be convenient.

This is just a personal opinion but instead of `/* */` comments, `//` can be more convenient at times. For example, when you need to comment a whole block of code. Watching the code base, although there are `/* */` comments there are not thaaat many.

Reply:
@knzo25 
Added an error message in latest commit

Reply:
Thanks !

Reply:
@Mingyu1991 
Just a suggestion. When addressing a comment with a commit, it is easier to check when you paste the hash in the same thread as the comment